File: Computational_Physics_Problem_Solving_with_Python_Landau/HPCodes/laplace.f
c  From "COMPUTATIONAL PHYSICS", 3rd Ed, Enlarged Python eTextBook  
c  by RH Landau, MJ Paez, and CC Bordeianu
c  Copyright Wiley-VCH Verlag GmbH & Co. KGaA, Berlin;  Copyright R Landau,
c  Oregon State Unv, MJ Paez, Univ Antioquia, C Bordeianu, Univ Bucharest, 2015.
c  Support by National Science Foundation
         
c  laplace.f: Solution of Laplace equation with finite differences 
c             Output in gnuplot 3D grid format used by

      Program laplace
			
      	Implicit none
      	Integer max
      	Parameter(max = 40)
      	Real*8 x, p(max, max)
      	Integer i, j, iter, y 
				
      	Open(8, File = 'laplace.dat', Status = 'Unknown')
c                                        side with constant potential
   	    Do 10 i = 1, max
   	      p(i, 1) = 100.
 10 	  Continue   
c                                                 iteration algorithm
  	    Do 20 iter = 1, 1000
  	      Do 30 i = 2, (max-1)
  	        Do 40 j = 2, (max-1)
  	    p(i, j) = 0.25*(p(i+1,j)+p(i-1,j) +p(i,j+1) + p(i,j-1))
 40	        Continue
 30	      Continue
 20	    Continue	    
c                                         output in gnuplot 3D format
	      Do 50 i = 1, max
	        Do 60 j = 1, max
	          Write (8, 22) p(i, j)
 60	      Continue
 	        Write (8, 22)
 50	    Continue
 22 	  Format(f10.6)
 	      Close(8)   
 	      Stop 'data saved in laplace.dat'
  	  End

File: Computational_Physics_Problem_Solving_with_Python_Landau/HPCodes/laplace.c
/* From "COMPUTATIONAL PHYSICS", 3rd Ed, Enlarged Python eTextBook  
    by RH Landau, MJ Paez, and CC Bordeianu
    Copyright Wiley-VCH Verlag GmbH & Co. KGaA, Berlin;  Copyright R Landau,
    Oregon State Unv, MJ Paez, Univ Antioquia, C Bordeianu, Univ Bucharest, 2015.
    Support by National Science Foundation
   
 laplace.c                          
*  comment: Output data is saved in 3D grid format used by gnuplot     *
************************************************************************
*/
#include <stdio.h>

#define max 40                         /* number of grid points */

main()
{
   double x, p[max][max];
   int i, j, iter, y;
   
   FILE *output;      /* save data in laplace.dat */
   output = fopen("laplace.dat","w");
   
   for(i=0; i<max; i++)                 /* clear the array  */
   {   
      for (j=0; j<max; j++) p[i][j] = 0;
   }

   for(i=0; i<max; i++) p[i][0] = 100.0;        /* p[i][0] = 100 V */   

   for(iter=0; iter<1000; iter++)               /* iterations */
   {
      for(i=1; i<(max-1); i++)                  /* x-direction */
      {
         for(j=1; j<(max-1); j++)               /* y-direction */
         {
            p[i][j] = 0.25*(p[i+1][j]+p[i-1][j]+p[i][j+1]+p[i][j-1]);
         }
      }
   }
   
   for (i=0; i<max ; i++)         /* write data gnuplot 3D format */
   {
      for (j=0; j<max; j++) 
      {
         fprintf(output, "%f\n",p[i][j]);
      }
      fprintf(output, "\n");    /* empty line for gnuplot */
   }
   printf("data stored in laplace.dat\n");
   fclose(output);
}
 
